quinn/src/connection.rs

```diff
@@ -857,16 +857,19 @@ impl State {
         let mut transmits = 0;
 
         let max_datagrams = self.socket.max_transmit_segments();
+        let capacity = self.inner.current_mtu();
+        let mut buffer = BytesMut::with_capacity(capacity as usize);
 
-        while let Some(t) = self.inner.poll_transmit(now, max_datagrams) {
+        while let Some(t) = self.inner.poll_transmit(now, max_datagrams, &mut buffer) {
             transmits += match t.segment_size {
                 None => 1,
-                Some(s) => (t.contents.len() + s - 1) / s, // round up
+                Some(s) => (t.size + s - 1) / s, // round up
             };
             // If the endpoint driver is gone, noop.
-            let _ = self
-                .endpoint_events
-                .send((self.handle, EndpointEvent::Transmit(t)));
+            let _ = self.endpoint_events.send((
+                self.handle,
+                EndpointEvent::Transmit(t, buffer.clone().freeze()),
```

Comment:
`BytesMut::clone` always performs a deep copy. We should use [`split_to`](https://docs.rs/bytes/latest/bytes/struct.BytesMut.html#method.split_to) instead to have any possibility of reduced allocs/copies. That will also make `clear` below unnecessary.